R/chat.R

```diff
@@ -535,8 +541,127 @@ print.ChatOpenAI <- function(x, ...) {
   invisible(x)
 }
 
+#' @export
+#' @examplesIf elmer:::openai_key_exists()
+#' chat <- new_chat_openai(echo = TRUE)
+#' chat$chat(
+#'   list(
+#'     "What do you see in these images?",
+#'     content_image_url("https://www.r-project.org/Rlogo.png"),
+#'     content_image_file(system.file("httr2.png", package = "elmer"))
+#'   )
+#' )
+content_image_url <- function(url, detail = c("auto", "low", "high")) {
+  # TODO: Allow vector input?
+  check_string(url)
+
+  list(type = "image_url", image_url = list(url = url))
+}
+
+#' @rdname content_image_url
+#' @export
+content_image_file <- function(path, detail = c("auto", "low", "high"), content_type = "auto", resize = NULL) {
+  # TODO: Allow vector input?
+  check_string(path)
+  if (!file.exists(path)) {
+    stop("File does not exist: ", path)
+  }
+
+  if (content_type == "auto") {
+    # OpenAI supports .png, .jpeg, .jpg, .webp, .gif
+    # https://platform.openai.com/docs/guides/vision/what-type-of-files-can-i-upload
+    ext <- tolower(tools::file_ext(path))
+    content_type <- switch(
+      ext,
+      png = "image/png",
+      jpeg = "image/jpeg",
+      jpg = "image/jpeg",
+      webp = "image/webp",
+      gif = "image/gif",
+      stop("Unsupported image file extension: ", ext)
+    )
+  }
+
+  if (!is.null(resize)) {
+    # Load the image
+    img <- magick::image_read(path)
+    img <- magick::image_scale(img, "512x512")
+    buf <- magick::image_write(img, format = magick::image_info(img)$format)
+    base64 <- base64enc::base64encode(buf)
+  } else {
+    base64 <- base64enc::base64encode(path)
+  }
+
+  data_uri <- paste0("data:", content_type, ";base64,", base64)
+
+  content_image_url(data_uri)
+}
+
 
 last_message <- function(chat) {
   messages <- chat$messages()
   messages[[length(messages)]]
 }
+
+# Define allowed types - add new types here in the future
+allowed_input_types <- c("text", "image_url")
+
+normalize_chat_input <- function(input) {
```

Comment:
If you want to make this full tidyverse style you'd pass `error_call = caller_env()` along so that error messages can include the user facing function call.

Reply:
Sorry, I tried but could not figure this out. I skimmed https://rlang.r-lib.org/reference/local_error_call.html and then added `error_call = caller_env()` to both `normalize_chat_input` and `process_single_input`, and also added `local_error_call(error_call)` to the top of both function bodies, but after doing that the errors would print with no context at all, instead of the caller as I expected.